### PR TITLE
Updated BPS example

### PR DESCRIPTION
**Pull request by mschauer**

### CODE REVIEW

File: benchmarks/bouncy.jl [hunk 1]
```diff
@@ -16,35 +15,29 @@ Random.seed!(1)
...
 end
-x, y = readlrdata()
+A, y = readlrdata()
 
-xt = collect(x')
+At = collect(A')
 
-m = @model (xt, y, σ) begin
-    d,n = size(xt)
+model = @model (At, y, σ) begin
```
Comment (cscherrer) [inline benchmarks/bouncy.jl:29]: This shadows `Tilde.model`, but I guess that's mostly for internal use anyway

File: benchmarks/bouncy.jl [hunk 2]
```diff
@@ -64,14 +57,21 @@ end
...
-d = 1 + 24 # number of parameters 
+d = 25 # number of parameters 
 t0 = 0.0
 x0 = zeros(d) # starting point sampler
-T = 500. # end time (similar to number of samples in MCMC)
-c = 5.0 # initial guess for the bound
-#M = I
-M = Diagonal(1 ./ [1.7, 0.08, 0.01, 0.09, 0.01, 0.06, 0.08, 0.12, 0.09, 0.11, 0.01, 0.11, 0.18, 0.29, 0.21, 0.88, 0.21, 0.39, 0.44, 0.65, 0.4, 0.35, 0.6, 0.31, 0.3])
+# estimated posterior mean
+xhat = [3.412, -0.5916, 0.03527, -0.3873, 0.004466, -0.2345, -0.15, -0.2164, 0.01229, 0.1739, -0.009764, -0.3217, 0.2173, 0.08189, -0.2851, -1.59, 0.6661, -1.003, 1.078, 1.401, 0.3271, -0.1361, -0.6437, -0.06795, -0.05282]
```
Comment (cscherrer) [inline benchmarks/bouncy.jl:64]: Where does this come from?

File: benchmarks/bouncy.jl [hunk 3]
```diff
@@ -81,26 +81,22 @@ Z = BouncyParticle(∅, ∅, # ignored
...
+using TupleVectors: chainvec
+using MeasureTheory: transform
+
 
 function collect_sampler(t, sampler, n)
     x1 = transform(t, sampler.u0[2][1])
     tv = chainvec(x1, n)
     ϕ = iterate(sampler)
     j = 1
+    global state
```
Comment (cscherrer) [inline benchmarks/bouncy.jl:99]: Why is this global?

**Review by cscherrer — COMMENTED** [review]
Just some comments/questions

File: benchmarks/bouncy.jl (hunk 2 above)
Comment (mschauer) [reply benchmarks/bouncy.jl:64]: A long run ;-) Still too far off, running for an hour or so would be good.

File: benchmarks/bouncy.jl (hunk 3 above)
Comment (mschauer) [reply benchmarks/bouncy.jl:99]:
```suggestion
```

File: benchmarks/bouncy.jl [hunk 4]
```diff
@@ -81,26 +81,22 @@ Z = BouncyParticle(∅, ∅, # ignored
...
-while ϕ !== nothing
-    val, state = ϕ
-    val[1]> 1 && break
-    println(val[1])
-    ϕ = iterate(sampler, state)
-end
-end
+
+using TupleVectors: chainvec
+using MeasureTheory: transform
```
Comment (mschauer) [inline benchmarks/bouncy.jl:91]:
```suggestion
using Tilde.MeasureTheory: transform
```

Comment (mschauer) [reply benchmarks/bouncy.jl:91]: Found while running on a fresh real

File: benchmarks/bouncy.jl (hunk 1 above)
Comment (mschauer) [reply benchmarks/bouncy.jl:29]: `lr` would be a good name

File: benchmarks/bouncy.jl [hunk 5]
```diff
@@ -109,49 +104,10 @@ function collect_sampler(t, sampler, n)
...
-
-
+tv = @time collect_sampler(as(post), sampler, 1000)
 
-# bps_chain = MCMCChains.Chains([xj[i] for xj in x[end÷4:end], i in 1:d])
-# bps_chain = setinfo(bps_chain,  (;start_time=0.0, stop_time = elapsed_time))
-# bps_chain
+using MCMCChains
+bps_chain = MCMCChains.Chains(tv.θ)
+bps_chain = setinfo(bps_chain,  (;start_time=0.0, stop_time = elapsed_time))
```
[the remark this answers is not in the document]
Comment (cscherrer) [reply benchmarks/bouncy.jl:111]: Oh wait, that's for the fixed-time version, iterative will be different

[the remark this answers is not in the document]
Comment (mschauer) [reply benchmarks/bouncy.jl:114]: Fixed measuring the iterator

**Comment by mschauer** [comment]
Should work now after updating your ZZB

File: benchmarks/bouncy.jl (hunk 2 above)
Comment (cscherrer) [reply benchmarks/bouncy.jl:65]: Could be a could test for the iteration interface. Have an `OnlineStat` to get the mean and its standard error, then iterate until it's small enough. Nice that it's O(1) memory :)

Comment (mschauer) [reply benchmarks/bouncy.jl]: The values are now from 
```

Iterations        = 1:1:100000
Number of chains  = 1
Samples per chain = 100000
Wall duration     = 797.86 seconds
Compute duration  = 797.86 seconds
parameters        = param_1, param_2, param_3, param_4, param_5, param_6, param_7, param_8, param_9, param_10, param_11, param_12, param_13, param_14, param_15, param_16, param_17, param_18, param_19, param_20, param_21, param_22, param_23, param_24, param_25

Summary Statistics
  parameters      mean       std   naive_se      mcse          ess      rhat   ess_per_sec 
      Symbol   Float64   Float64    Float64   Float64      Float64   Float64       Float64 

     param_1    3.4061    1.2161     0.0038    0.0070   13524.1959    1.0000       16.9505
     param_2   -0.5918    0.0718     0.0002    0.0003   35007.5577    1.0000       43.8766
     param_3    0.0352    0.0087     0.0000    0.0000   41978.5689    1.0000       52.6137
     param_4   -0.3874    0.0873     0.0003    0.0004   28044.5615    1.0000       35.1496
     param_5    0.0045    0.0039     0.0000    0.0000   42075.2679    1.0000       52.7349
     param_6   -0.2346    0.0600     0.0002    0.0002   51872.1793    1.0000       65.0138
     param_7   -0.1495    0.0768     0.0002    0.0003   31132.7961    1.0001       39.0202
     param_8   -0.2184    0.1160     0.0004    0.0004   45446.2455    1.0001       56.9599
     param_9    0.0122    0.0835     0.0003    0.0003   33343.5490    1.0000       41.7910
    param_10    0.1731    0.1002     0.0003    0.0004   36290.9562    1.0001       45.4851
    param_11   -0.0098    0.0086     0.0000    0.0000   51873.4303    1.0000       65.0154
    param_12   -0.3224    0.1120     0.0004    0.0004   52825.1399    1.0000       66.2082
    param_13    0.2168    0.1630     0.0005    0.0007   19602.8532    1.0000       24.5692
    param_14    0.0800    0.2368     0.0007    0.0008   31890.6219    1.0000       39.9700
    param_15   -0.2829    0.1947     0.0006    0.0007   45101.0133    1.0001       56.5272
    param_16   -1.5810    0.6373     0.0020    0.0040   13396.6897    1.0000       16.7907
    param_17    0.6666    0.1970     0.0006    0.0008   23346.0037    1.0000       29.2606
    param_18   -0.9984    0.3503     0.0011    0.0016   18501.4790    1.0000       23.1888
    param_19    1.0808    0.4261     0.0013    0.0015   40923.7440    1.0000       51.2916
    param_20    1.4054    0.5699     0.0018    0.0021   48434.0404    1.0000       60.7046
    param_21    0.3270    0.3683     0.0012    0.0017   20234.2626    1.0000       25.3605
    param_22   -0.1357    0.3245     0.0010    0.0015   17202.6026    1.0000       21.5608
    param_23   -0.6446    0.6180     0.0020    0.0025   30793.6932    1.0000       38.5952
    param_24   -0.0658    0.3256     0.0010    0.0015   25712.1416    1.0000       32.2262
    param_25   -0.0499    0.2641     0.0008    0.0011   23706.7103    1.0000       29.7127
```

**Comment by mschauer** [comment]
@sethaxen Pathfinder, Tilde and ZigZagBoomerang nicely working as team :-) We think of comparing times with some other things out there

**Comment by sethaxen** [comment]
> @sethaxen Pathfinder, Tilde and ZigZagBoomerang nicely working as team :-) We think of comparing times with some other things out there

Nice! Anything you need on the Pathfinder side?

**Comment by mschauer** [comment]
```
struct LWrapper{T}
    Σ::T
end
M = LWrapper(pf_result.fit_distribution.Σ)
x0 = pf_result.fit_distribution.μ;
using Pathfinder.PDMats
function ZigZagBoomerang.reflect!(∇ϕx, x, θ, F::BouncyParticle)
    θ .-= (2*dot(∇ϕx, θ)/PDMats.quad(F.L.Σ, ∇ϕx))*(F.L.Σ*∇ϕx)
    θ
end
function ZigZagBoomerang.refresh!(rng, θ, F::BouncyParticle)
    ρ̄ = sqrt(1-F.ρ^2)
    θ .*= F.ρ
    u = ρ̄*PDMats.unwhiten(F.L.Σ, randn(rng, length(θ)))
    θ .+= u
    θ
end
```
Hot-Patch to allow to use woodbury decomposition given by Pathfinder,  https://github.com/sethaxen/Pathfinder.jl/issues/77

**Comment by sethaxen** [comment]
> Hot-Patch to allow to use woodbury decomposition given by Pathfinder, [sethaxen/Pathfinder.jl#77](https://github.com/sethaxen/Pathfinder.jl/issues/77)

Very nice! You can eek out a little more performance with this implementation:
```julia
function ZigZagBoomerang.reflect!(∇ϕx, x, θ, F::BouncyParticle)
    z = F.L.Σ * ∇ϕx
    θ .-= (2*dot(∇ϕx, θ)/dot(∇ϕx, z)) * z
    θ
end
```

**Comment by sethaxen** [comment]
@mschauer what is `LWrapper` meant to represent? Is it $L$, such that $L L^\top = \Sigma^{-1}$?

**Comment by mschauer** [comment]
Yeah, but I am not sure it’s a very sensible design right now, I have to look how dynamichmc etc deal with metrics

**Comment by mschauer** [comment]
Are we happy with the design from ZZB side for this use case? Then I'll "freeze" that into a type while adding support for metrics.

**Comment by mschauer** [comment]
Yes, so I would like to make a version BouncyParticle without these inconsistency, so I would like to know if just fixing these two issues would do

**Comment by sethaxen** [comment]
> Very nice! You can eek out a little more performance with this implementation:

I'm pretty sure we can also implement a non-allocating `dot(::AbstractVecOrMat, ::WoodburyPDMat, ::AbstractVecOrMat)` method, but it's only worth doing so if `F.L.Σ * ∇ϕx` is a computational bottleneck.

**Comment by cscherrer** [comment]
Thanks @mschauer . Since the transform guarantees we're in the support, I think we should be using `unsafe_logdensityof` for this case

**Comment by mschauer** [comment]
I am done for now, with introduction of PDMats support, for now I will not refactor the `struct BouncyParticle`.

File: benchmarks/bouncy.jl [hunk 6]
```diff
@@ -1,157 +1,165 @@
...
 # define BouncyParticle sampler (has two relevant parameters) 
-Z = BouncyParticle(∅, ∅, # ignored
-    2.0, # momentum refreshment rate 
+Z = BouncyParticle(missing, # Graphical structure 
+    MAP, # ignored
+    2.0, # momentum refreshment rate and sample saving rate 
     0.95, # momentum correlation / only gradually change momentum in refreshment/momentum update
-    0.0, # ignored
-    M # cholesky of momentum precision
+    M, # metric
```
Comment (mschauer) [inline benchmarks/bouncy.jl:93]:
```suggestion
    M, # metric (PDMat compatible object for momentum covariance)
```

File: benchmarks/bouncy.jl [hunk 7]
```diff
@@ -1,157 +1,165 @@
...
+    ∅
```
Comment (mschauer) [inline benchmarks/bouncy.jl:94]:
```suggestion
    missing # legacy
```

**Pull request merged**